### PR TITLE
[api] Rename `global_reference` to `GlobRef.t` to follow kernel style.

### PR DESCRIPTION
**Pull request by ejgallego**
In #6092, `global_reference` was moved to `kernel`. It makes sense to
go further and use the current kernel style for names.

This has a good effect on the dependency graph, as some core modules
don't depend on library anymore.

A question about providing equality for the GloRef module remains, as
there are two different notions of equality for constants. In that
sense, `KerPair` seems suspicious and at some point it should be
looked at.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
globnames also has the refordered/set/map stuff, should that move to names?
I don't know what the pair stuff is about.

**Comment by ejgallego** [comment]
> globnames also has the refordered/set/map stuff, should that move to names?

I didn't want to move it yet as not to increase more the trusted base, but I think we maybe should, along with the comparsion.

> I don't know what the pair stuff is about.

See `Names.KerPair` basically constants and inductive posses two orderings, which makes it difficult to provide a usable module.

**Review by herbelin — COMMENTED** [review]
In itself, I like what this PR does, and, actually, I had hesitated for long at some time at whether the `Ind`/`Const`/`Construct` node of `constr` should not be replaced by a `GlobRef` node. I was afraid of the cost of adapting the code, and I did not do it.

Unfortunately, I did not make a lot of progresses in my thinking since then. I approve the content PR but what matters more in my opinion is, even w/o going to the extreme of a `Constr.GlobRef` node, whether a consensus can be obtained in the direction of the cleaning.

Said otherwise, I feel that we, or at least a lot of us, feel (much) more comfortable with the directions taken by the ongoing restructuration of the code, and that this will make further extensions easier, but we shall have to use our best argumentation to justify that plugin writers have to adapt to the changes.

Remark that I'm actually speculating a lot since I don't have a clear picture of who develop / maintain plugins in general.

In any case, I would be more interested in a clear statement from everyone of us. For instance, why @ppedrot stayed in the middle of the path and did not move all `Term` to the `Constr` model?

**Comment by ejgallego** [comment]
Hi Hugo, thanks for the comments,



> we shall have to use our best argumentation to justify that plugin writers have to adapt to the changes.
> 
> Remark that I'm actually speculating a lot since I don't have a clear picture of who develop / maintain plugins in general.

This is a very good question; I think that indeed that more discussion about the plugins is needed and I reserved a slow at the NextWG to try to have some more discussion.

On the other hand adapting to the changes, even if tedious, is not so painful due to OCaml's deprecation mechanism. As a plugin writer, if you fix all the warnings you should be almost ready for the next Coq version.

> In any case, I would be more interested in a clear statement from everyone of us. For instance, why @ppedrot stayed in the middle of the path and did not move all Term to the Constr model?

Indeed good question, I guess just a bit more of work may be needed to clean that area up.

> In itself, I like what this PR does, and, actually, I had hesitated for long at some time at whether the Ind/Const/Construct node of constr should not be replaced by a GlobRef node. I was afraid of the cost of adapting the code, and I did not do it.

Matthieu also pointed out that having this in kernel would be an advantage, and due the use of this type in many interfaces IMHO the change made sense. On the other hand, I dunno about propagating this `Constr`, it could indeed have some performance impact due to extra allocation.

**Comment by ejgallego** [comment]
Only the lint fails in Travis, and that, as in other renaming PRs, is expected.

**Comment by herbelin** [comment]
> On the other hand, I dunno about propagating this Constr, it could indeed have some performance impact due to extra allocation.

It can also provide more sharing, so it is not clear to me that it would be more allocation at the end.

**Comment by SkySkimmer** [comment]
>In itself, I like what this PR does, and, actually, I had hesitated for long at some time at whether the Ind/Const/Construct node of constr should not be replaced by a GlobRef node. I was afraid of the cost of adapting the code, and I did not do it.

I notice that this doesn't include the Var node. In #6092 I also do not use the VarRef node (except for writing assert false cases where needed). Does this maybe indicate that what we want in the kernel is a type with only Const/Ind/Construct?

**Comment by ejgallego** [comment]
> Does this maybe indicate that what we want in the kernel is a type with only Const/Ind/Construct?

Interesting. I guess we need to have a look to the users, for example equations uses `global_reference` quite a bit.

**Comment by SkySkimmer** [comment]
Also I don't know much about the module code but if we ever allow instantiating parameters by inductive types / constructors I imagine that would involve mod_subst having a function of type like `substitution -> globref -> globref`, but ignoring vars.

**Comment by ejgallego** [comment]
> Also I don't know much about the module code but if we ever allow instantiating parameters by inductive types / constructors I imagine that would involve mod_subst having a function of type like substitution -> globref -> globref, but ignoring vars.

You mean moving the current `Globnames.subst_*` functions to `Mod_subst`?

**Comment by SkySkimmer** [comment]
Maybe.
Whatever turns `M.T` into `I.T` in this:
```coq
Module Type T. Parameter T : Type. End T.

Module F (M:T). Definition T := M.T. End F.

Module I. Definition T := nat. End I.

Module A := F I.
Print A.T. (* A.T = I.T
     : Type *)
```

**Comment by maximedenes** [comment]
@ejgallego Is this ready to merge or do you want to take more of @SkySkimmer 's remarks into account?

**Comment by mattam82** [comment]
For plugins I suppose they generally use the non-var subset of global_reference (to refer to defined constants/inductives) but VarRef almost always makes sense in tactics where we want to refer indifferently to a defined entry or to a local goal variable or section variable. The kernel itself need only handle Var right now not VarRefs though, so we could indeed have two types to reflect this.

**Comment by ejgallego** [comment]
I think the remarks can be taken care in additional PRs, however I do wonder about the remarks from @mattam82 and @herbelin . I am not sure if they hare happy with this going in as-is [and later refining] or they would like to further refine before merging.

**Comment by ejgallego** [comment]
I'm holding up the rebase until we discuss a bit more about what the plan should be here.

**Comment by ejgallego** [comment]
A few comments:

- should this include `Var` maybe yes? what are the kernel users? `my_global_ref`? do all the potential kernel users don't use the `Var` case?
- what should the notion of equality implemented here?
- Hugo's comment about refactoring `Constr` seems very interesting.

**Comment by maximedenes** [comment]
@ppedrot and @gares think refactoring constr would be too costly performance-wise

**Comment by ejgallego** [comment]
> @ppedrot and @gares think refactoring constr would be too costly performance-wise

Well, it is not so clear, some cases will indeed get strongly penalized, but others can do better.

**Comment by ejgallego** [comment]
So with #6092 merged this PR is just a cleanup on the way to deprecate `globnames`.

It'd be great to have this deprecation in 8.8 but yeah, I understand we won't.

**Comment by ejgallego** [comment]
Ok so this is ready to go in I think [modulo CI of course].

I really dunno what to do for 8.8 as in a sense #6092 should have depended on this one.

**Comment by SkySkimmer** [comment]
>#6092 should have depended on this one.

Why?

**Comment by ejgallego** [comment]
> Why?

Because the only thing this PR does is to move `global_reference` to `Names`, so it seems logical that the move is done and reviewed separately.

**Comment by ejgallego** [comment]
In particular introducing an alias without deprecation is a bit inconvenient. That being said, I'd just merge this and forget about the thing.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
Wrt the XXX and the `GlobRefUser` I think @maximedenes  can tell you which is which in user/canonical

Wrt the proposal of @herbelin , I can probably bring some experience:
- In Matita 0.5 we had exactly the same terms of Coq
- In Matita 0.9 we have a single node `Const of Uri.t` where `Uri.t = Constant .. | InductiveType .. | Constructor`. It works well for writing the kernel. The advantage is that a lot of code treats all global names the same way (they have a type in the env, etc). You pay the price for an extra level of pattern matching in reduction & co, but you can also cache in the `Uri.t` some piece of info (`Uri.t` is an hashconsed datatype in Matita). For example `Constructor` had the number of parameters as an argument: to reduce a match you don't need to look up in the env how many arguments have to be dropped (the parameters) in order to beta reduce the function corresponding to the branch.
- In elpi I tried a mixed approach: I have the `const | indt | indc` nodes (`Var` is just an artifact, I map it to `const` for section vars or to a bound variable for proof variables) but they all have a `GlobRef.t` as an argument (that is an opaque data type in elpi). This means that I can have an API `type-of-gr` that gives me the type or any global constant, instead of 3 API. At the same time some code has 3 identical cases. You gain a little in performance (indexing in elpi goes down 1 level, so dispatching is faster with 3 nodes that differ at the top level). I hope that for a "user" having 3 nodes is more sensible than having just one (semantically the 3 objects are very different, and the fact that in CIC you have just 1 syntax for everything is a big source of confusion to newcomers). The "UI" of elpi has always been a concern to me, since it targets users, not developers.  In a way I've opted for a compromise here. It does not work too bad, but I don't love it either (but I'm a dev).

For Coq proper, I believe that the approach of Matita 0.9 would be best, but it is an enormous change.

My 2 C

**Review by forestjulien — APPROVED** [review]

File: kernel/names.mli
Comment (maximedenes) [inline kernel/names.mli]: I see that  `eq_gr` uses canonical names. I find that a bit surprising. We should probably review the uses to know if equality modulo user names makes sense.

**Review by maximedenes — APPROVED** [review]

File: kernel/names.mli
Comment (ejgallego) [reply kernel/names.mli]: Yup.

**Comment by SkySkimmer** [comment]
@ejgallego why does this have 2 assignees?

**Comment by ejgallego** [comment]
> @ejgallego why does this have 2 assignees?

How many devs does it take to merge a PR?

**Comment by maximedenes** [comment]
> How many devs does it take to merge a PR?

I suggest you follow the documented process instead of doing assignments yourself.

**Comment by maximedenes** [comment]
This doesn't compile on my machine because of a deprecated warning. See `exception ClearDependencyError` in `evarutil.mli`. Was probably added after this PR.

**Comment by ejgallego** [comment]
> I suggest you follow the documented process instead of doing assignments yourself.

Hey light-bulb-derived jokes are always good. BTW I assigned this before I fully understood the new docs/merging process.

> This doesn't compile on my machine because of a deprecated warning. See exception ClearDependencyError in evarutil.mli. Was probably added after this PR.

Strange, it does compile here in both forms, rebased anyways.

**Comment by maximedenes** [comment]
> Strange, it does compile here in both forms, rebased anyways.

Wasn't the deprecated alias used in the exception I mentioned?

**Comment by ejgallego** [comment]
Rebased.

**Review by PierreCorbineau — APPROVED** [review]

**Comment by maximedenes** [comment]
Still getting a warning on Travis. Is this spurious?

File: engine/evarutil.ml
Comment (SkySkimmer) [inline engine/evarutil.ml]: This change needs to be reflected in the mli.

**Comment by ejgallego** [comment]
Ahh, I got bitten by the race condition that means that you don't recompile a file twice even if a warning was emitted, thus my build compiled... fixed.

**Comment by SkySkimmer** [comment]
Sounds like you don't user -warn-error.

**Review by forestjulien — APPROVED** [review]

**Comment by ejgallego** [comment]
This [since long time] trivial patch has seen close to **200** rebases, some days I wonder what I am doing with my life. Pierre-Marie, would you care taking care of it? Merci.

**Comment by maximedenes** [comment]
This is a kernel patch, so I'm the assignee. FTR I've been tracking this PR for 5 days only, due to a wrong assignment. Since then, it's been ready for 2 days only, then waiting for CI. Don't hesitate to ping me when CI is complete, if you want things to get merged quickly (that is, during working hours).

**Comment by ejgallego** [comment]
@maximedenes failure in gitlab is spurious, so this is ready CI-wise, cheers and thanks for looking into this.

**Review by maximedenes — APPROVED** [review]

**Pull request merged**
